// ngx-http-loader/esm2015/ngx-http-loader.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { NgxHttpLoaderService, NgxHttpLoaderComponent, NgxHttpLoaderModule, NgxLoader } from './public_api';
export { ClockLoaderComponent as ɵf } from './lib/loaders/clock-loader/clock-loader.component';
export { DefaultLoaderComponent as ɵg } from './lib/loaders/default-loader/default-loader.component';
export { DualRingLoaderComponent as ɵe } from './lib/loaders/dual-ring-loader/dual-ring-loader.component';
export { EllipsisLoaderComponent as ɵh } from './lib/loaders/ellipsis-loader/ellipsis-loader.component';
export { FacebookLoaderComponent as ɵi } from './lib/loaders/facebook-loader/facebook-loader.component';
export { GearLoaderComponent as ɵc } from './lib/loaders/gear-loader/gear-loader.component';
export { GridLoaderComponent as ɵd } from './lib/loaders/grid-loader/grid-loader.component';
export { HeartLoaderComponent as ɵj } from './lib/loaders/heart-loader/heart-loader.component';
export { HourglassLoaderComponent as ɵk } from './lib/loaders/hourglass-loader/hourglass-loader.component';
export { JellyboxLoaderComponent as ɵs } from './lib/loaders/jellybox-loader/jellybox-loader.component';
export { MulticircleLoaderComponent as ɵr } from './lib/loaders/multicircle-loader/multicircle-loader.component';
export { RingLoaderComponent as ɵl } from './lib/loaders/ring-loader/ring-loader.component';
export { RippleLoaderComponent as ɵm } from './lib/loaders/ripple-loader/ripple-loader.component';
export { RollerLoaderComponent as ɵn } from './lib/loaders/roller-loader/roller-loader.component';
export { RotatingPlaneComponent as ɵt } from './lib/loaders/rotating-plane/rotating-plane.component';
export { SpinnerLoaderComponent as ɵo } from './lib/loaders/spinner-loader/spinner-loader.component';
export { ThreeBounceLoaderComponent as ɵq } from './lib/loaders/three-bounce-loader/three-bounce-loader.component';
export { WaterWaveLoaderComponent as ɵp } from './lib/loaders/water-wave-loader/water-wave-loader.component';
export { NgxHttpLoaderInterceptor as ɵa, RequestsInterceptorProvider as ɵb } from './lib/ngx-loader-interceptor.service';
