// ngx-capture/esm2015/lib/ngx-capture.service.js
import { Injectable } from '@angular/core';
import html2canvas from 'html2canvas';
import { from, of } from 'rxjs';
import * as i0 from "@angular/core";
export class NgxCaptureService {
    getImage(screen, fullCapture, cropDimensions) {
        let options = {
            logging: false,
        };
        if (!fullCapture && cropDimensions.width > 10 && cropDimensions.height > 10) {
            options = Object.assign(Object.assign({}, options), cropDimensions);
        }
        else if (!fullCapture) {
            return of(null);
        }
        return from(html2canvas(screen, options)
            .then((canv) => {
            const img = canv.toDataURL('image/png');
            return img;
        }, (err) => {
            throw new Error(err);
        })
            .catch((res) => {
            throw new Error(res);
        }));
    }
    // TODO
    downloadImage(img) {
        // this.canvas.nativeElement.src = img;
        // this.downloadLink.nativeElement.href = img;
        // this.downloadLink.nativeElement.download = 'test.png';
        // this.downloadLink.nativeElement.click();
    }
}
NgxCaptureService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgxCaptureService_Factory() { return new NgxCaptureService(); }, token: NgxCaptureService, providedIn: "root" });
NgxCaptureService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
