// ngx-http-loader/esm5/lib/ngx-loader-interceptor.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { finalize } from 'rxjs/operators';
import * as i0 from "@angular/core";
var NgxHttpLoaderInterceptor = /** @class */ (function () {
    function NgxHttpLoaderInterceptor() {
        this.reqStatus$ = new ReplaySubject(1);
        this.pendingReqCnt = 0;
        this.forceStop = false;
        this.filterUrls = [];
        this.filterUrlPatterns = [];
        this.filterMethods = [];
        this.filterHeaders = [];
    }
    Object.defineProperty(NgxHttpLoaderInterceptor.prototype, "pendingRequestCnt", {
        get: /**
         * @return {?}
         */
        function () {
            return this.pendingRequestCnt;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgxHttpLoaderInterceptor.prototype, "reqStatus", {
        get: /**
         * @return {?}
         */
        function () {
            return this.reqStatus$.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgxHttpLoaderInterceptor.prototype, "setFilterUrls", {
        set: /**
         * @param {?} urls
         * @return {?}
         */
        function (urls) {
            this.filterUrls = urls;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgxHttpLoaderInterceptor.prototype, "setUrlPatterns", {
        set: /**
         * @param {?} regx
         * @return {?}
         */
        function (regx) {
            this.filterUrlPatterns = regx;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgxHttpLoaderInterceptor.prototype, "setFilterMethods", {
        set: /**
         * @param {?} methods
         * @return {?}
         */
        function (methods) {
            this.filterMethods = methods;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgxHttpLoaderInterceptor.prototype, "setFilterHeaders", {
        set: /**
         * @param {?} headers
         * @return {?}
         */
        function (headers) {
            this.filterHeaders = headers;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgxHttpLoaderInterceptor.prototype, "setForceStopReq", {
        set: /**
         * @param {?} bool
         * @return {?}
         */
        function (bool) {
            this.forceStop = bool;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} url
     * @return {?}
     */
    NgxHttpLoaderInterceptor.prototype.checkIsFilterUrl = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        return this.filterUrls.some(function (x) { return x === url; });
    };
    /**
     * @param {?} url
     * @return {?}
     */
    NgxHttpLoaderInterceptor.prototype.checkUrlPatterns = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        return this.filterUrlPatterns.some(function (x) { return x.test(url); });
    };
    /**
     * @param {?} req
     * @return {?}
     */
    NgxHttpLoaderInterceptor.prototype.checkHttpMethod = /**
     * @param {?} req
     * @return {?}
     */
    function (req) {
        return this.filterMethods.some(function (e) {
            return e.toLowerCase() === req.method.toLowerCase();
        });
    };
    /**
     * @param {?} req
     * @return {?}
     */
    NgxHttpLoaderInterceptor.prototype.checkReqHeader = /**
     * @param {?} req
     * @return {?}
     */
    function (req) {
        return this.filterHeaders.some(function (e) {
            return req.headers.has(e);
        });
    };
    /**
     * @param {?} r
     * @return {?}
     */
    NgxHttpLoaderInterceptor.prototype.isStopReq = /**
     * @param {?} r
     * @return {?}
     */
    function (r) {
        return this.forceStop
            || this.checkIsFilterUrl(r.urlWithParams)
            || this.checkUrlPatterns(r.urlWithParams)
            || this.checkHttpMethod(r)
            || this.checkReqHeader(r);
    };
    /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    NgxHttpLoaderInterceptor.prototype.intercept = /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    function (request, next) {
        var _this = this;
        /** @type {?} */
        var isNext = this.isStopReq(request);
        if (!isNext) {
            this.pendingReqCnt++;
            if (1 === this.pendingReqCnt) {
                this.reqStatus$.next(true);
            }
        }
        return next.handle(request).pipe(finalize(function () {
            if (!isNext) {
                _this.pendingReqCnt--;
                if (0 === _this.pendingReqCnt) {
                    _this.reqStatus$.next(false);
                }
            }
        }));
    };
    NgxHttpLoaderInterceptor.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ NgxHttpLoaderInterceptor.ngInjectableDef = i0.defineInjectable({ factory: function NgxHttpLoaderInterceptor_Factory() { return new NgxHttpLoaderInterceptor(); }, token: NgxHttpLoaderInterceptor, providedIn: "root" });
    return NgxHttpLoaderInterceptor;
}());
export { NgxHttpLoaderInterceptor };
if (false) {
    /** @type {?} */
    NgxHttpLoaderInterceptor.prototype.reqStatus$;
    /** @type {?} */
    NgxHttpLoaderInterceptor.prototype.pendingReqCnt;
    /** @type {?} */
    NgxHttpLoaderInterceptor.prototype.forceStop;
    /** @type {?} */
    NgxHttpLoaderInterceptor.prototype.filterUrls;
    /** @type {?} */
    NgxHttpLoaderInterceptor.prototype.filterUrlPatterns;
    /** @type {?} */
    NgxHttpLoaderInterceptor.prototype.filterMethods;
    /** @type {?} */
    NgxHttpLoaderInterceptor.prototype.filterHeaders;
}
/** @type {?} */
export var RequestsInterceptorProvider = [{
        provide: HTTP_INTERCEPTORS,
        useExisting: NgxHttpLoaderInterceptor,
        multi: true
    }];
