// ngx-http-loader/esm5/lib/ngx-http-loader.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { NgxHttpLoaderInterceptor } from './ngx-loader-interceptor.service';
import * as i0 from "@angular/core";
import * as i1 from "./ngx-loader-interceptor.service";
var NgxHttpLoaderService = /** @class */ (function () {
    function NgxHttpLoaderService(reqService) {
        this.reqService = reqService;
        this.showLoader = new ReplaySubject(1);
    }
    Object.defineProperty(NgxHttpLoaderService.prototype, "isShowLoader", {
        get: /**
         * @return {?}
         */
        function () {
            return this.showLoader.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NgxHttpLoaderService.prototype.show = /**
     * @return {?}
     */
    function () {
        this.reqService.setForceStopReq = true;
        this.showLoader.next(true);
    };
    /**
     * @return {?}
     */
    NgxHttpLoaderService.prototype.hide = /**
     * @return {?}
     */
    function () {
        this.showLoader.next(false);
        this.reqService.setForceStopReq = false;
    };
    NgxHttpLoaderService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    NgxHttpLoaderService.ctorParameters = function () { return [
        { type: NgxHttpLoaderInterceptor }
    ]; };
    /** @nocollapse */ NgxHttpLoaderService.ngInjectableDef = i0.defineInjectable({ factory: function NgxHttpLoaderService_Factory() { return new NgxHttpLoaderService(i0.inject(i1.NgxHttpLoaderInterceptor)); }, token: NgxHttpLoaderService, providedIn: "root" });
    return NgxHttpLoaderService;
}());
export { NgxHttpLoaderService };
if (false) {
    /** @type {?} */
    NgxHttpLoaderService.prototype.showLoader;
    /** @type {?} */
    NgxHttpLoaderService.prototype.reqService;
}
