// ngx-http-loader/esm2015/lib/ngx-loader-interceptor.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { finalize } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class NgxHttpLoaderInterceptor {
    constructor() {
        this.reqStatus$ = new ReplaySubject(1);
        this.pendingReqCnt = 0;
        this.forceStop = false;
        this.filterUrls = [];
        this.filterUrlPatterns = [];
        this.filterMethods = [];
        this.filterHeaders = [];
    }
    /**
     * @return {?}
     */
    get pendingRequestCnt() {
        return this.pendingRequestCnt;
    }
    /**
     * @return {?}
     */
    get reqStatus() {
        return this.reqStatus$.asObservable();
    }
    /**
     * @param {?} urls
     * @return {?}
     */
    set setFilterUrls(urls) {
        this.filterUrls = urls;
    }
    /**
     * @param {?} regx
     * @return {?}
     */
    set setUrlPatterns(regx) {
        this.filterUrlPatterns = regx;
    }
    /**
     * @param {?} methods
     * @return {?}
     */
    set setFilterMethods(methods) {
        this.filterMethods = methods;
    }
    /**
     * @param {?} headers
     * @return {?}
     */
    set setFilterHeaders(headers) {
        this.filterHeaders = headers;
    }
    /**
     * @param {?} bool
     * @return {?}
     */
    set setForceStopReq(bool) {
        this.forceStop = bool;
    }
    /**
     * @param {?} url
     * @return {?}
     */
    checkIsFilterUrl(url) {
        return this.filterUrls.some(x => x === url);
    }
    /**
     * @param {?} url
     * @return {?}
     */
    checkUrlPatterns(url) {
        return this.filterUrlPatterns.some(x => x.test(url));
    }
    /**
     * @param {?} req
     * @return {?}
     */
    checkHttpMethod(req) {
        return this.filterMethods.some(e => {
            return e.toLowerCase() === req.method.toLowerCase();
        });
    }
    /**
     * @param {?} req
     * @return {?}
     */
    checkReqHeader(req) {
        return this.filterHeaders.some(e => {
            return req.headers.has(e);
        });
    }
    /**
     * @param {?} r
     * @return {?}
     */
    isStopReq(r) {
        return this.forceStop
            || this.checkIsFilterUrl(r.urlWithParams)
            || this.checkUrlPatterns(r.urlWithParams)
            || this.checkHttpMethod(r)
            || this.checkReqHeader(r);
    }
    /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    intercept(request, next) {
        /** @type {?} */
        const isNext = this.isStopReq(request);
        if (!isNext) {
            this.pendingReqCnt++;
            if (1 === this.pendingReqCnt) {
                this.reqStatus$.next(true);
            }
        }
        return next.handle(request).pipe(finalize(() => {
            if (!isNext) {
                this.pendingReqCnt--;
                if (0 === this.pendingReqCnt) {
                    this.reqStatus$.next(false);
                }
            }
        }));
    }
}
NgxHttpLoaderInterceptor.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ NgxHttpLoaderInterceptor.ngInjectableDef = i0.defineInjectable({ factory: function NgxHttpLoaderInterceptor_Factory() { return new NgxHttpLoaderInterceptor(); }, token: NgxHttpLoaderInterceptor, providedIn: "root" });
if (false) {
    /** @type {?} */
    NgxHttpLoaderInterceptor.prototype.reqStatus$;
    /** @type {?} */
    NgxHttpLoaderInterceptor.prototype.pendingReqCnt;
    /** @type {?} */
    NgxHttpLoaderInterceptor.prototype.forceStop;
    /** @type {?} */
    NgxHttpLoaderInterceptor.prototype.filterUrls;
    /** @type {?} */
    NgxHttpLoaderInterceptor.prototype.filterUrlPatterns;
    /** @type {?} */
    NgxHttpLoaderInterceptor.prototype.filterMethods;
    /** @type {?} */
    NgxHttpLoaderInterceptor.prototype.filterHeaders;
}
/** @type {?} */
export const RequestsInterceptorProvider = [{
        provide: HTTP_INTERCEPTORS,
        useExisting: NgxHttpLoaderInterceptor,
        multi: true
    }];
