// ngx-http-loader/esm2015/lib/ngx-http-loader.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, TemplateRef } from '@angular/core';
import { merge, timer } from 'rxjs';
import { switchMap, debounce, distinctUntilChanged, tap, partition } from 'rxjs/operators';
import { NgxHttpLoaderInterceptor } from './ngx-loader-interceptor.service';
import { NgxLoader } from './ngx-http-loader';
import { NgxHttpLoaderService } from './ngx-http-loader.service';
import { DomSanitizer } from '@angular/platform-browser';
export class NgxHttpLoaderComponent {
    /**
     * @param {?} requestService
     * @param {?} service
     * @param {?} sanitizer
     */
    constructor(requestService, service, sanitizer) {
        this.requestService = requestService;
        this.service = service;
        this.sanitizer = sanitizer;
        this.customEnumLoader = NgxLoader;
        this.loader = NgxLoader.RIPPLE;
        this.filterUrlRegexPatterns = [];
        this.filterUrls = [];
        this.filterHttpHeaders = [];
        this.filterHttpMethods = [];
        this.delay = 0;
        this.extraDelay = 0;
        this.minDuration = 0;
        this.opacity = 0.8;
        // @Input() public backdropBackgroundColor = 'linear-gradient(45deg, #222428, #444856)';
        // @Input() public backdropBackgroundColor = 'transperent';
        this.backdropBackgroundColor = 'linear-gradient(0deg, black, transparent);';
        this.iconColor = 'white';
        this.title = '';
        this.titleColor = '#fff';
        this.titleSize = '3rem';
        this.subTitle = '';
        this.subTitleColor = '#6c757d';
        this.subTitleSize = '1.2rem';
        this.custom = false;
        this.showuntil = Date.now();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (this.filterUrlRegexPatterns.length > 0 || this.filterUrls.length > 0
            || this.filterHttpHeaders.length > 0 || this.filterHttpMethods.length > 0)
            this.initAllFilters();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.initAllFilters();
    }
    /**
     * @return {?}
     */
    initAllFilters() {
        this.requestService.setUrlPatterns = this.filterUrlRegexPatterns;
        this.requestService.setFilterUrls = this.filterUrls;
        this.requestService.setFilterMethods = this.filterHttpMethods;
        this.requestService.setFilterHeaders = this.filterHttpHeaders;
        this.initShowObservable();
    }
    /**
     * @return {?}
     */
    initShowObservable() {
        const [show, hide] = partition(($hide) => $hide)(this.requestService.reqStatus);
        this.show$ = merge(this.requestService.reqStatus.pipe(switchMap(() => show.pipe(debounce(() => timer(this.delay))))), show.pipe(switchMap(() => hide.pipe(debounce(() => this.showUpto())))), this.service.isShowLoader).pipe(distinctUntilChanged(), tap(x => this.updateHideDelay(x)));
    }
    /**
     * @param {?} bool
     * @return {?}
     */
    updateHideDelay(bool) {
        if (bool)
            this.showuntil = Date.now() + this.minDuration;
    }
    /**
     * @return {?}
     */
    showUpto() {
        return timer(Math.max(this.extraDelay, this.showuntil - Date.now()));
    }
}
NgxHttpLoaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-http-loader',
                template: "<div *ngIf=\"show$ | async\">\r\n    <ng-template [ngTemplateOutlet]=\"customTemplate || defaultTemplate\"></ng-template>\r\n    <ng-template #defaultTemplate>\r\n        <div class=\"custom-loader vertical-center\"\r\n            [attr.style]=\"sanitizer.bypassSecurityTrustStyle('--background-color: ' + backdropBackgroundColor + '; opacity :' + opacity)\">\r\n            <div class=\"w-100 text-center text-light\"\r\n                [attr.style]=\"sanitizer.bypassSecurityTrustStyle('--iconcolor: ' + iconColor)\">\r\n                <ng-container *ngIf=\"custom\">\r\n                    <ng-template [ngTemplateOutlet]=\"customLoader\"></ng-template>\r\n                </ng-container>\r\n                <ng-container *ngIf=\"!custom\">\r\n                    <lib-gear-loader *ngIf=\"loader === customEnumLoader.GEAR\"></lib-gear-loader>\r\n                    <lib-grid-loader *ngIf=\"loader === customEnumLoader.GRID\"></lib-grid-loader>\r\n                    <lib-dual-ring-loader *ngIf=\"loader === customEnumLoader.DUALRING\"></lib-dual-ring-loader>\r\n                    <lib-clock-loader *ngIf=\"loader === customEnumLoader.CLOCK\"></lib-clock-loader>\r\n                    <lib-default-loader *ngIf=\"loader === customEnumLoader.DEFAULT\"></lib-default-loader>\r\n                    <lib-ellipsis-loader *ngIf=\"loader === customEnumLoader.ELLIPSIS\"></lib-ellipsis-loader>\r\n                    <lib-spinner-loader *ngIf=\"loader === customEnumLoader.SPINNER\"></lib-spinner-loader>\r\n                    <lib-facebook-loader *ngIf=\"loader === customEnumLoader.FACEBOOK\"></lib-facebook-loader>\r\n                    <lib-heart-loader *ngIf=\"loader === customEnumLoader.HEART\"></lib-heart-loader>\r\n                    <lib-hourglass-loader *ngIf=\"loader === customEnumLoader.HOURGLASS\"></lib-hourglass-loader>\r\n                    <lib-ring-loader *ngIf=\"loader === customEnumLoader.RING\"></lib-ring-loader>\r\n                    <lib-ripple-loader *ngIf=\"loader === customEnumLoader.RIPPLE\"></lib-ripple-loader>\r\n                    <lib-roller-loader *ngIf=\"loader === customEnumLoader.ROLLER\"></lib-roller-loader>\r\n                    <lib-water-wave-loader *ngIf=\"loader === customEnumLoader.WATERWAVE\"></lib-water-wave-loader>\r\n                    <lib-multicircle-loader *ngIf=\"loader === customEnumLoader.MULTICIRCLE\"></lib-multicircle-loader>\r\n                    <lib-three-bounce-loader *ngIf=\"loader === customEnumLoader.THREEBOUNCE\"></lib-three-bounce-loader>\r\n                    <lib-jellybox-loader *ngIf=\"loader === customEnumLoader.JELLYBOX\"></lib-jellybox-loader>\r\n                    <lib-rotating-plane *ngIf=\"loader === customEnumLoader.ROTATINGPLANE\"></lib-rotating-plane>\r\n                </ng-container>\r\n                <h1 *ngIf=\"title\" class=\"font-weight-bold text-shadow\"\r\n                    style=\"font-family:Trebuchet MS,Helvetica,sans-serif; width: 50%; margin: 10px auto; margin-top: 20px; word-break: break-all;\" \r\n                    [style.color]=\"titleColor\" [style.fontSize]=\"titleSize\">\r\n                    {{title}}\r\n                </h1>\r\n                <div style=\"width: 50%; margin: 10px auto; word-break: break-all;\" *ngIf=\"subTitle\" [style.color]=\"subTitleColor\" [style.fontSize]=\"subTitleSize\">{{subTitle}}</div>\r\n            </div>\r\n        </div>\r\n    </ng-template>\r\n</div>",
                styles: [".custom-loader{position:fixed;top:0;left:0;width:100%;height:100%;background:var(--background-color);color:#fff;z-index:99999;transition:.3s linear}.custom-loader ::ng-deep>*{vertical-align:middle;white-space:normal;display:inline-block;z-index:99999999999}.vertical-center:before{content:\"\";display:inline-block;vertical-align:middle;width:0;height:100%}"]
            }] }
];
/** @nocollapse */
NgxHttpLoaderComponent.ctorParameters = () => [
    { type: NgxHttpLoaderInterceptor },
    { type: NgxHttpLoaderService },
    { type: DomSanitizer }
];
NgxHttpLoaderComponent.propDecorators = {
    customTemplate: [{ type: Input }],
    customLoader: [{ type: Input }],
    loader: [{ type: Input }],
    filterUrlRegexPatterns: [{ type: Input }],
    filterUrls: [{ type: Input }],
    filterHttpHeaders: [{ type: Input }],
    filterHttpMethods: [{ type: Input }],
    delay: [{ type: Input }],
    extraDelay: [{ type: Input }],
    minDuration: [{ type: Input }],
    opacity: [{ type: Input }],
    backdropBackgroundColor: [{ type: Input }],
    iconColor: [{ type: Input }],
    title: [{ type: Input }],
    titleColor: [{ type: Input }],
    titleSize: [{ type: Input }],
    subTitle: [{ type: Input }],
    subTitleColor: [{ type: Input }],
    subTitleSize: [{ type: Input }],
    custom: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.customEnumLoader;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.customTemplate;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.customLoader;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.loader;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.filterUrlRegexPatterns;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.filterUrls;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.filterHttpHeaders;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.filterHttpMethods;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.delay;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.extraDelay;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.minDuration;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.opacity;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.backdropBackgroundColor;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.iconColor;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.title;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.titleColor;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.titleSize;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.subTitle;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.subTitleColor;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.subTitleSize;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.custom;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.show$;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.showuntil;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.requestService;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.service;
    /** @type {?} */
    NgxHttpLoaderComponent.prototype.sanitizer;
}
