// ngx-http-loader/esm2015/lib/ngx-http-loader.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { NgxHttpLoaderInterceptor } from './ngx-loader-interceptor.service';
import * as i0 from "@angular/core";
import * as i1 from "./ngx-loader-interceptor.service";
export class NgxHttpLoaderService {
    /**
     * @param {?} reqService
     */
    constructor(reqService) {
        this.reqService = reqService;
        this.showLoader = new ReplaySubject(1);
    }
    /**
     * @return {?}
     */
    get isShowLoader() {
        return this.showLoader.asObservable();
    }
    /**
     * @return {?}
     */
    show() {
        this.reqService.setForceStopReq = true;
        this.showLoader.next(true);
    }
    /**
     * @return {?}
     */
    hide() {
        this.showLoader.next(false);
        this.reqService.setForceStopReq = false;
    }
}
NgxHttpLoaderService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
NgxHttpLoaderService.ctorParameters = () => [
    { type: NgxHttpLoaderInterceptor }
];
/** @nocollapse */ NgxHttpLoaderService.ngInjectableDef = i0.defineInjectable({ factory: function NgxHttpLoaderService_Factory() { return new NgxHttpLoaderService(i0.inject(i1.NgxHttpLoaderInterceptor)); }, token: NgxHttpLoaderService, providedIn: "root" });
if (false) {
    /** @type {?} */
    NgxHttpLoaderService.prototype.showLoader;
    /** @type {?} */
    NgxHttpLoaderService.prototype.reqService;
}
